multiple cards in passport

ichub opened the pull request:
closes: https://github.com/proofcarryingdata/zupass/issues/98
closes: https://github.com/proofcarryingdata/zupass/issues/100
closes: https://github.com/proofcarryingdata/zupass/issues/183
closes: https://github.com/proofcarryingdata/zupass/issues/184
follow-up item: https://github.com/proofcarryingdata/zupass/issues/182
follow-up item: https://github.com/proofcarryingdata/zupass/issues/187
follow-up item: https://github.com/proofcarryingdata/zupass/issues/188

https://user-images.githubusercontent.com/2636237/232631884-6331b0e9-bf57-4020-80c4-f921f431168e.mov

dcposch commented:
I think the cards could use a name and description. What does the card below mean to a user? What can they do with it? If they have two of them, how do they tell them apart?

<img width="416" alt="image" src="https://user-images.githubusercontent.com/169280/232245625-d0c1502f-56a5-4bc3-b606-99ace60c81c9.png">

dcposch commented on apps/consumer-client/pages/examples/add-pcd.tsx:
```diff
@@ -0,0 +1,53 @@
+          add a new semaphore identity to the passport
+        </button>
+        {/* TODO: implement a response which indicates whether the PCD was
+        added successfully or not */}
```
👍

dcposch commented on apps/consumer-client/pages/examples/prove-and-add.tsx:
```diff
@@ -0,0 +1,101 @@
+    PASSPORT_URL,
+    window.location.origin + "/popup",
+    SemaphoreGroupPCDPackage.name,
+    {
```
can this JSON object live in semaphoregrouppcd? (or at least the field names and types. then, the caller could specify a partial object if they want some of them to be not user-provided.) feels like a lot to copy around

ichub replied:
I personally kinda prefer it to be explicitly defined every time so the dev does have to think about it.

dcposch commented on apps/passport-client/components/screens/AddScreen/AddScreen.tsx:
```diff
@@ -0,0 +1,30 @@
+import { JustAddScreen } from "./JustAddScreen";
+import { ProveAndAddScreen } from "./ProveAndAddScreen";
+
+export function AddScreen() {
```
```
/** Asks the user to add a new PCD card to their passport. */
```

dcposch commented on apps/passport-client/components/screens/AddScreen/AddScreen.tsx:
```diff
@@ -0,0 +1,30 @@
+    return <JustAddScreen request={request as PCDAddRequest} />;
+  }
+
+  err(dispatch, "Unsupported request", `Expected a PCD ADD request`);
```
👍

dcposch commented on apps/passport-client/components/screens/HomeScreen.tsx:
```diff
@@ -29,11 +26,13 @@ export function HomeScreen() {
+  const zuzaluPCDId = useMemo(() => {
+    return state.pcds.getAll()[0].id;
+  }, [state.pcds]);
+  const [selectedPCDID, setSelectedPCDID] = useState(() => zuzaluPCDId);
```
nit but this can be just `= useState(zuzaluID)`

& i'd be consistent (`PCDID` vs `PCDId` => maybe just `ID`?)

dcposch commented on apps/passport-client/components/screens/HomeScreen.tsx:
```diff
@@ -29,11 +26,13 @@ export function HomeScreen() {
     }
   });
 
```
may be worth adding
```ts
const pcds = useMemo(() => state.pcds.getAll(), [state.pcds]);
```
...to reduce duplication

dcposch commented on apps/passport-client/components/screens/HomeScreen.tsx:
```diff
@@ -43,24 +42,24 @@ export function HomeScreen() {
+                expanded={pcd.id === selectedPCD.id}
+                isZuzaluIdentity={pcd.id === zuzaluPCDId}
+                onClick={() => {
+                  console.log("CLICKED");
```
delete

dcposch commented on apps/passport-client/components/screens/ProveScreen/GenericProveScreen.tsx:
```diff
@@ -23,84 +22,42 @@ import { PCDArgs } from "../../shared/PCDArgs";
+        onProve={onProve}
+        pcdType={req.pcdType}
+        options={req.options}
+      />
```
👍

dcposch commented on apps/passport-client/components/screens/ProveScreen/GenericProveSection.tsx:
```diff
@@ -0,0 +1,100 @@
+import { RippleLoader } from "../../core/RippleLoader";
+import { PCDArgs } from "../../shared/PCDArgs";
+
+export function GenericProveSection<T extends PCDPackage = PCDPackage>({
```
short comment explaining what the GenericProveSection does

dcposch commented on apps/passport-client/components/screens/VerifyScreen.tsx:
```diff
@@ -89,9 +93,7 @@ export function VerifyScreen() {
-          {result?.valid === true && (
-            <CardElem expanded card={getCard(result)} />
-          )}
+          {result && getCard(result)}
```
this will throw an error if `result.valid === false`, making the message above redundant

dcposch commented on apps/passport-client/components/shared/PCDCard.tsx:
```diff
@@ -0,0 +1,180 @@
+    }
+  }
+
+  console.log("");
```
delete

dcposch commented on apps/passport-client/components/shared/PCDCard.tsx:
```diff
@@ -0,0 +1,180 @@
+    <>
+      <TextCenter>
+        {pcd.type} <br />
+        no implementation of a ui for this type of card found
```
`{pcd.type} unsupported`, or maybe show the error screen

dcposch commented on apps/passport-client/src/dispatch.ts:
```diff
@@ -233,6 +240,22 @@ function resetPassport() {
+    await savePCDs(state.pcds);
+    update({ pcds: state.pcds });
+  } else {
+    console.log("ERROR");
```
`"Can't add PCD, unsupported type " + pcd.type`

dcposch replied:
and def throw an error rather than just printing

dcposch commented on packages/passport-interface/src/PassportInterface.ts:
```diff
@@ -52,38 +66,33 @@ export function constructPassportPcdGetRequestUrl<T extends PCDPackage>(
+    options,
+  };
+  const eqReq = encodeURIComponent(JSON.stringify(req));
+  return `${passportOrigin}#/add?request=${eqReq}`;
```
👍

dcposch commented on packages/passport-interface/src/PassportPopup.ts:
```diff
@@ -66,6 +67,10 @@ export function usePassportPopupSetup() {
+    } else if (pcdResponse != null) {
+      // Later, the Passport redirects back with a encodedPendingPCD. Send it to our parent.
+      window.opener.postMessage({ response: pcdResponse }, "*");
+      window.close();
```
👍

dcposch commented on apps/passport-client/components/shared/PCDCard.tsx:
```diff
@@ -0,0 +1,180 @@
+  }, [pcd, pcdPackage]);
+  const header = isZuzaluIdentity
+    ? "VERIFIED ZUZALU PASSPORT"
+    : displayOptions?.header ?? "PCD";
```
if you separated out `PCDCard` from `ZuzaluCard`, do you still need the special case?

& for typographic consistency, the header should be `toUpperCase()`

ichub replied:
Yes. Under the hood, a zuzalu card is a semaphore identity pcd, so I am special casing it to make it behave differently from other semaphore identity cards.

ichub commented:
> ### I think the cards could use a name and description.
> 
> What does the card below mean to a user? What can they do with it? If they have two of them, how do they tell them apart?
> 
> <img alt="image" width="416" src="https://user-images.githubusercontent.com/169280/232245625-d0c1502f-56a5-4bc3-b606-99ace60c81c9.png">

<img width="459" alt="Screenshot 2023-04-17 at 10 07 57 PM" src="https://user-images.githubusercontent.com/2636237/232598599-2a62424e-e29d-4e4a-b20f-1d56fa7ee7c8.png">

OK, so the way i've implemented it each PCD is responsible for rendering its own card. I've added descriptions for each of the existing cards.

I'm not sure about how to go about having each card have a name. My current plan is to give each PCD Package a function that given a PCD returns a name, which will be mostly useful for disambiguating PCDs in the dropdown in the generic prove screen.
